### ISSUE TITLE
Typing dynamic required arguments 

### ISSUE THREAD

**Issue by hulla-dev** [facebook/flow#8455]
I'm having little trouble flowtype*ing* this correctly. Basically I'm passing an object of arguments where I'm internally checking if the required arguments were passed. Issue is, I'm not sure how to infer this to flow

Consider the following typing:

```js
// @flow
export type DownloadUrlParams = {|
  businessId?: string,
  jobId?: string,
  from?: Date,
  to?: Date,
|}
```

The reason why it's so strangely typed, is because these are optional arguments, that can be required in various combinations.

```js
// @flow
export const PATHS = {
   reports: {
      required: ['businessId']
      // $ExpectError
      get: ({ businessId }) => `business/${businessId}/reports`
   },
   // ... etc, leaving out the rest for sake of example simplicity
}
```

and then finally inside my function

```js
// @flow
import * as R from 'ramda'

const getDownloadUrl = (path: $Keys<typeof PATHS>, params: DownloadUrlParams) => {
   const pathArgs = R.pick(PATHS[path].required, params)
   // ^ gets .required object properties, basically { ...[required] } = object

   const hasAllArgs = (paramsObj: DownloadUrlParams) => R.toPairs(paramsObj)
      .every((entry) => {
        // R.toPairs === Object.entries
        if (!entry[1]) {
          throw Error(`Fetch type: ${path}. Missing required param ${entry[0]}`)
        }
        return !!entry[1]
      }

   hasAllArgs(pathArgs)
   return PATHS[path].get(pathArgs)
}
```

Thing is, I know the arguments provided are valid, because the `hasAllArgs()` function would throw the error otherwise, I just need to somehow convey this information to the [tag:flowtype]

I tried to enforce cast a type for the object I'm passing:

```js
type PickedObject = {|
   [$Keys<DownloadUrlParams>]: $Values<DownloadUrlParams>
|}

type EnforceTypeFunction = <V>(value: V) => $NonMaybeType<V>
type RequiredArgs = $ObjMap<PickedObject, EnforceTypeFunction>

const pathArgs: RequiredArgs = { ...R.pick(PATHS[path].required, params) }
```

This somewhat works and is seemingly step in the right direction, issue is, now the parameter itself is on either of those two types, instead of the exact type per parameter

[![enter image description here][1]][1]
[![enter image description here][2]][2]

basically i need to somehow enforce instead of

```js
[key: string]: Date | string
```

to

```js
[key: string]: // <exact type of the specific value>
```

  [1]: https://i.stack.imgur.com/iRTCD.png
  [2]: https://i.stack.imgur.com/A3Pbu.png

 - **Link**: 🔗 https://stackoverflow.com/questions/63279660/typing-required-picked-arguments

**Comment by SamChou19815** [comment]
I think now you can use conditional type for this
